### PR TITLE
Support multiple audit boards and multi-county audits

### PR DESCRIPTION
**Pull request by gws**
This is it! We now support multiple audit boards auditing simultaneously. Much more code had to change than you might have thought.

**Further work:**

Subsequent round testing is needed to make sure the audit board behavior across rounds works as intended. I've been extensively testing on a single round with some numbers of phantom ballots. Otherwise, the code makes very few modifications to the machinery of the audit itself, which is desirable as that's about to change later this week when multi-county audit work lands.

**References:**

https://www.pivotaltracker.com/epic/show/3921089

**Support Multi-County Audits**

Also incorporated on top of the branch used by this PR is #37 "Support Multi-County Audits", which has a nice Overview section on how they work, the concept of Ballot Universes, and the important fact that "Every contest is a multi-county contest" now.

### CODE REVIEW

File: client/src/selector/county/canAudit.ts [hunk 1]
```diff
@@ -2,8 +2,13 @@ import auditBoardSignedIn from './auditBoardSignedIn';
 
 
 function canAudit(state: County.AppState) {
+    return state.asm.county === 'COUNTY_AUDIT_UNDERWAY';
+
+    // TODO: Do we need to test auditBoardSignedIn?
+    /*
```
Comment (gws) [inline client/src/selector/county/canAudit.ts]: :spiral_notepad: I need to revisit this, I think this was an early comment.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRToAuditDownload.java [hunk 2]
```diff
@@ -234,10 +235,36 @@ public String endpointBody(final Request the_request, final Response the_respons
...
+          // each response row.
+          for (int i = 0; i < bsa.size(); i++) {
+            final Map<String, Integer> m = bsa.get(i);
+
+            final Integer boardIndex = m.get("index");
+            final Integer boardCount = m.get("count");
+
+            for (int j = boardIndex; j < boardIndex + boardCount; j++) {
+              // TODO: Will this always agree with the round information?
+              final CVRToAuditResponse row = response_list.get(j);
```
Comment (gws) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRToAuditDownload.java]: :spiral_notepad: Tentatively the answer is "yes", but deserves scrutiny.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/model/Round.java [hunk 3]
```diff
@@ -278,7 +303,74 @@ public void setActualAuditedPrefixLength(final int the_audited_prefix_length) {
...
+    this.ballotSequenceAssignment = l;
+  }
+
+  /**
+   * Returns the list of CVRs under audit in this round.
+   *
+   * @return a list whose indices correspond to audit board indices and values
+   *         being the next CVR for the given audit board to audit.
+   */
+  // TODO: Extract into query class
```
Comment (gws) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/model/Round.java]: :spiral_notepad: This is still possible.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRToAuditDownload.java (hunk 2 above)
Comment (sduckett) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRToAuditDownload.java]: :thinking:

File: server/eclipse-project/src/main/java/us/freeandfair/corla/controller/BallotSelection.java [hunk 4]
```diff
@@ -23,13 +23,20 @@
...
-      in "audit sequence order"
-   **/
+  /**
+   * Select CVRs associated with the ballots in the manifest, if they exist.
+   *
+   * CVRs are returned in an order corresponding to the order in rands.
+   *
+   * @param rands random numbers corresponding to ballot manifest positions
+   * @param countyId the county ID
+   */
```
Comment (sduckett) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/controller/BallotSelection.java]: :flags: Watch for conflicts here when combining with multi-county.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/model/CountyDashboard.java [hunk 5]
```diff
@@ -484,24 +548,28 @@ public void startRound(final int the_number_of_ballots,
...
+   * the round.
+   *
    * @exception IllegalStateException if there is no current round.
    */
-  public void endRound(final List<Elector> the_signatories) {
+  public void endRound() {
     if (my_current_round_index == null) {
       throw new IllegalStateException("no round to end");
     } else {
+      this.setAuditBoardCount(null);
```
Comment (teaforthecat) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/model/CountyDashboard.java]: Will the County Admin have to set the AuditBoardCount on the next round? Why not keep it around?

Comment (gws) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/model/CountyDashboard.java]: The thought was that the county might be able to send audit boards home.

Comment (gws) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/model/CountyDashboard.java]: We need to test round transitions a bit better, maybe we can revisit this, but the actual setting of the audit board count is what triggers the code to map ballots onto boards as well.

**Pull request merged**